From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Mon, 15 Jun 2026 22:41:15 +0800
Subject: [PATCH 41/47] experiments: cpu_pstate_latencies: Support older ADTs
 by using cpu.reg fallback

Support older ADTs by using cpu.reg instead of cpu.cpu_id when cpu_id property
is not available. cpu.cpu_id is only possibly unavailable on systems with a
single cluster type where bit [17] of MPIDR is 0, so this approximation is
sufficient.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 proxyclient/experiments/cpu_pstate_latencies.py | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/proxyclient/experiments/cpu_pstate_latencies.py b/proxyclient/experiments/cpu_pstate_latencies.py
index ffd86a8f..53b0c121 100755
--- a/proxyclient/experiments/cpu_pstate_latencies.py
+++ b/proxyclient/experiments/cpu_pstate_latencies.py
@@ -14,11 +14,18 @@ tfreq = u.mrs(CNTFRQ_EL0)
 
 TEST_CPUS = []
 CLUSTER_TYPES = []
+
+def get_cpuid(cpu):
+    if hasattr(cpu, "cpu_id"):
+        return cpu.cpu_id
+    else:
+        return cpu.reg
+
 for cpu in u.adt["/cpus"]:
-    if cpu.cpu_id == 0 or cpu.state == 'running':
+    if get_cpuid(cpu) == 0 or cpu.state == 'running':
         continue
     if not hasattr(cpu, "cluster_type"):
-        TEST_CPUS.append(cpu.cpu_id)
+        TEST_CPUS.append(get_cpuid(cpu))
         CLUSTER_TYPES.append("")
         break
     elif cpu.cluster_type not in CLUSTER_TYPES:
